### ISSUE TITLE
android/qgroundcontrol auto record tlog in a directory under /sdcard not working

### ISSUE THREAD

**Issue by marcmerlin** [mavlink/qgroundcontrol#5276]
I believe qgroundcontrol has a way to manually save tlogs, but it should do so automatically with no user input.
I did a few flights with qgroundcontrol on android yesterday, and none of those flights got recorded on my phone. 
I have "enable automatic logging" in Mavlink 2.0 Logging (PX4 Firmware Only), but it didn't do anything. Lower down in 'saved log files', I have absolutely nothing.

Most of the time, sadly qgroundcontrol stopped running due to the phone going to sleep, or some other reason, so the main point is that it should automatically record all flights without user input until the app dies or gets terminated (which can be done at inconvenient times on android vs a desktop version).
While I do like qgroundcontrol's interface and extra functionality, for now I have to go back to Tower for the reliable auto track recording feature.
But I'm looking forward to going back to qgroundcontrol as soon as the track recording is reliable :)

**Comment by DonLakeFlyer** [comment]
@dogmaphobic I thought you had tuned on support for telemetry logs on Android but now on by default. But it look like the settings doesn't show up.

**Comment by dogmaphobic** [comment]
It "should" be there. I'll take a look when I get back.

**Comment by dogmaphobic** [comment]
I spent some time on this today and I can't quite tell what's going on. The _Start Log_ is properly sent and acknowledged, the log file is created but no log data is ever received. I'm using SITL as I don't have anything functional at the moment. After arming/disarming the "vehicle" several times, I ended up with several log files but obviously all with zero length as no data ever arrived.

**Comment by DonLakeFlyer** [comment]
If the setting is showing up and I can turn it on then I can debug from there.

**Comment by dogmaphobic** [comment]
It's all there. The same thing is happening on desktop. No log data arrives.

**Comment by dogmaphobic** [comment]
Does SITL support this?

**Comment by DonLakeFlyer** [comment]
Sure. Telemetry logs are just mavlink traffic.

**Comment by dogmaphobic** [comment]
No, this is not telemetry log. These are ulog logs (as described above).

**Comment by dogmaphobic** [comment]
>I have "enable automatic logging" in Mavlink 2.0 Logging (PX4 Firmware Only)

**Comment by DonLakeFlyer** [comment]
@marcmerlin I'm confused: You talk about saving tlogs which are telemetry logs. But you also talk about using "enable automatic logging" in Mavlink 2.0 Logging (PX4 Firmware Only)" which are vehicle logs, which are not telemetry logs. Do you want telemetry logs on mobile?

@dogmaphobic I was talking about saving telemetry logs on mobile which I though you had made available on mobile builds but I don't see the options showing up for that when I build android master.

@dogmaphobic With regards to the vehicle log automomatic logging stuff: Are you running that over a mavlink 2 connection? If you aren't I assume it will not work. Does the code check that it has a mavlink 2 connection before it tries to do anything? Also I would suggest changing the UI for this to better indicate that these are vehicle .ulogs (whatever the name for those are) and no telemetry logs.

**Comment by DonLakeFlyer** [comment]
I think we also need to look at app sleep again: https://github.com/mavlink/qgroundcontrol/issues/5354

**Comment by dogmaphobic** [comment]
Ok, this is as far as I got with telemetry log on Android (mobile in general). If logging is enabled, it is created and written to. Log is saved to a temp file and at the end, it is copied to the final location. On mobile, this never happens because the call that does all that is never called. Here is the function that closes (and eventually issues the signal to do the copy):

https://github.com/mavlink/qgroundcontrol/blob/master/src/comm/MAVLinkProtocol.cc#L330-L336

I disconnecting the vehicle (shutting down SITL). It took QGC about 20 seconds to figure out communication was lost. I proceeded to "Disconnect". In that case the log was properly saved.

In summary, logging is being written but never finalized unless you disconnect from the vehicle first. My guess is the application shutdown order on mobile is different and `_vehicleCountChanged()` is never called on exit.

Note that all this was only possible after forcibly enabling telemetry log, which is disabled by default on mobile and there is no UI to enable it.

![screen](https://user-images.githubusercontent.com/749243/27655904-09dcc00e-5c15-11e7-8746-37d477d23de1.png)

**Comment by marcmerlin** [comment]
@DonLakeFlyer I want/need to log tlog on my phone. 
I think you just explained to me that what I tried to turn on does not do what I actually need.
A GCS should have a way to log tlog for debugging or vehicle recovery if you have a crash in a remote location, ideally it should do so automatically (the other GCS I tried, do). 
Does qgroundcontrol allow this?

**Comment by DonLakeFlyer** [comment]
That seems like it is working correctly. The log is written when:
* Vehicle disconnects
* Or last vehicle disarms (depending on settings)

Since mobile has strange shutdown behavior you need to do one of the above to get it to close out. I think that is ok.

**Comment by DonLakeFlyer** [comment]
Given the mobile shutdown behavior the options to save logs even if vehicle is not armed may not be a good one. If you have everything else working could you put it up in a pull then I'll work with that to figure out what to do.

**Comment by DonLakeFlyer** [comment]
> Does qgroundcontrol allow this?

@marcmerlin Working on it.

**Comment by marcmerlin** [comment]
@DonLakeFlyer what you describe on writing the log (disconnect or disarm) sounds great, it's just not what I saw when I tried it.
Actually I got no log file at all of any of the flights I did.
Just to be clear, is it supposed to auto start saving of tlog under /sdcard somewhere? If so, where?
If it's not automatic, what is the user supposed to do to make it automatic?
Or your last update is saying that it's not doing this yet, but that you are working on it so that it will?

**Comment by DonLakeFlyer** [comment]
You couldn't have tried it since you can't currently turn it on from android. What you used was to download binary logs from the vehicle. That is a different thing.

**Comment by marcmerlin** [comment]
@DonLakeFlyer ok, that clears up the confusion, thanks.
I'll wait for a new version where this is turned on, thanks for your replies and looking at this.

**Comment by DonLakeFlyer** [comment]
Definitely something wrong with shutdown sequence on android

**Comment by DonLakeFlyer** [comment]
All working now. @marcmerlin Can you retry with tomorrow's android daily and let us know how it worked for you? Should be fully through the build system by then.

**Comment by marcmerlin** [comment]
@DonLakeFlyer thanks for doing this. Sadly, all 3 of my RC aircraft got destroyed last week (2 yesterday), so right now I'm unable to test anything :(

**Comment by DonLakeFlyer** [comment]
Yikes!
